### PR TITLE
use implied bounds compat mode in MIR borrowck

### PR DESCRIPTION
**Pull request by lcnr**
cc
- #119956
- #118553

This should hopefully fix bevy :thinking: `cargo test` ends up freezing my computer though, cargo build went from err to ok however :grin:

r? @jackh726

### CODE REVIEW

File: tests/ui/implied-bounds/normalization-nested-no-compat.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// Test for normalization of projections that appear in the item bounds
+// (versus those that appear directly in the input types).
+//
+// FIXME(-Zno-implied-bounds-compat): remove this test once this flag is removed,
+// duplicate of `normalization-nested.rs`.
```
Comment (jackh726) [inline tests/ui/implied-bounds/normalization-nested-no-compat.rs:5]: Can you just make 4 revisions in `normalization-nested`? lifetime or param x compat or non-comat?

File: tests/ui/implied-bounds/bevy_world_query.rs [hunk 2]
```diff
@@ -19,7 +19,9 @@ impl<Q: WorldQuery + 'static> SystemParam for Query<Q> {
 
 pub struct ParamSet<T: SystemParam>(T) where T::State: Sized;
 
-fn handler<'a>(_: ParamSet<Query<&'a u8>>) {}
+fn handler<'a>(x: ParamSet<Query<&'a u8>>) {
+    let _: ParamSet<_> = x;
```
Comment (jackh726) [inline tests/ui/implied-bounds/bevy_world_query.rs:23]: Is this enough to reproduce the bevy regression?

**Comment by jackh726** [comment]
@bors rollup=never p=1

r=me, preferably with the new test combined into the previous with revisions (and confirmation that the bevy regression is covered)

File: tests/ui/implied-bounds/bevy_world_query.rs (hunk 2 above)
Comment (lcnr) [reply tests/ui/implied-bounds/bevy_world_query.rs:23]: yes

Comment (jackh726) [reply tests/ui/implied-bounds/bevy_world_query.rs:23]: makes sense

**Comment by Kobzol** [comment]
Wins outweigh the losses, but `wg-grammar` has been behaving frantically lately, anyway.

@rustbot label: +perf-regression-triaged

**Pull request merged**